rustfmt-core/rustfmt-lib/src/items.rs

```diff
@@ -1897,39 +1903,42 @@ fn rewrite_static(
     }
 }
 
-pub(crate) fn rewrite_associated_type(
+pub(crate) fn rewrite_type_alias(
     ident: ast::Ident,
     ty_opt: Option<&ptr::P<ast::Ty>>,
     generics: &ast::Generics,
     generic_bounds_opt: Option<&ast::GenericBounds>,
     context: &RewriteContext<'_>,
     indent: Indent,
+    vis: &ast::Visibility,
 ) -> Option<String> {
-    let ident_str = rewrite_ident(context, ident);
-    // 5 = "type "
-    let generics_shape = Shape::indented(indent, context.config).offset_left(5)?;
-    let generics_str = rewrite_generics(context, ident_str, generics, generics_shape)?;
-    let prefix = format!("type {}", generics_str);
-
-    let type_bounds_str = if let Some(bounds) = generic_bounds_opt {
-        if bounds.is_empty() {
-            String::new()
-        } else {
-            // 2 = ": ".len()
-            let shape = Shape::indented(indent, context.config).offset_left(prefix.len() + 2)?;
-            bounds.rewrite(context, shape).map(|s| format!(": {}", s))?
-        }
-    } else {
-        String::new()
-    };
+    let mut prefix = rewrite_type_prefix(
+        context,
+        indent,
+        &format!("{}type ", format_visibility(context, vis)),
+        ident,
+        generics,
+        generic_bounds_opt,
+    )?;
 
     if let Some(ty) = ty_opt {
         // 1 = `;`
         let shape = Shape::indented(indent, context.config).sub_width(1)?;
-        let lhs = format!("{}{} =", prefix, type_bounds_str);
+
+        // If there's a where clause, add a newline before the assignment. Otherwise just add a
+        // space.
```

Comment:
Do we always have to add a newline and/or is this something we'd expect users to want to configure? I get the sense that this will be fairly subjective and some folks won't be fond of all the multilining for these type aliases

Reply:
I copied this from the existing implementation for writing types:

https://github.com/rust-lang/rustfmt/blob/12564f15fd9366db89000e88c432a5edafe2e927/rustfmt-core/rustfmt-lib/src/items.rs#L1599-L1623

Reply:
Yeah I figured, and I know it's consistent (both with existing type alias formatting and on functions/structs/enums). I suppose folks can also use `where_single_line` as well to control things a bit as well.

It just felt odd to me that an alias as short as `type A<'a> where 'a: 'static;` would be split into 4 lines by default (understood that will become 3 after the semi placement change discussed above)

Reply:
Yeah, I agree it looks a little odd. I think keeping consistency with how where clauses are formatted in other contexts (i.e. writing the where on a newline) is valuable. Do you find the 3-line version with the semicolon placement change to be nicer?

Reply:
Yeah, ideally, we respect the `where_single_line` option wherever we format a `where` clause. Unfortunately, that's not the case in the current implementation. I think that we can fix this as a part of the stabilization process of the `where_single_line` option; in this PR, we can follow the existing approach.

Reply:
SGTM :+1: